Crashing slaves cause a hard panic when a client is waiting on their commands

## Description of Problem
## System details
- **`uname -a`**: Darwin burke.local 12.2.0 Darwin Kernel Version 12.2.0: Sat Aug 25 00:48:52 PDT 2012; root:xnu-2050.18.24~1/RELEASE_X86_64 x86_64
- **`ruby -v`**: ruby 1.9.3p194 (2012-04-20 revision 35410) [x86_64-darwin12.2.0]
## Steps to Reproduce

1) `rails new foo ; cd foo`

2) `zeus start`

3) Before zeus finishes booting, run `zeus cucumber` in a new shell
## Observed Behavior

The master process hard panic's.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x75648]

goroutine 13 [running]:
github.com/burke/zeus/go/unixsocket.(*Usock).WriteMessage(0x0, 0xf8402b99e0, 0x3a430000000a, 0xf840000440, 0x8, ...)
        /Users/burke/go/src/github.com/burke/zeus/go/unixsocket/unixsocket.go:81 +0xba
github.com/burke/zeus/go/zeusmaster.(*SlaveNode).bootCommand(0xf840085340, 0xf8401455e0, 0xf8401455e0, 0x0)
        /Users/burke/go/src/github.com/burke/zeus/go/zeusmaster/slavenode.go:280 +0x9c
github.com/burke/zeus/go/zeusmaster.(*SlaveNode).bootQueuedCommandsAndSlaves(0xf840085340, 0x1)
        /Users/burke/go/src/github.com/burke/zeus/go/zeusmaster/slavenode.go:300 +0x7d
github.com/burke/zeus/go/zeusmaster.(*SlaveNode).doCrashedOrReadyState(0xf840085340, 0x1, 0x10f044, 0x4300000001)
        /Users/burke/go/src/github.com/burke/zeus/go/zeusmaster/slavenode.go:233 +0xd1
github.com/burke/zeus/go/zeusmaster.(*SlaveNode).Run(0xf840085340, 0xf840000540, 0x0, 0x0)
        /Users/burke/go/src/github.com/burke/zeus/go/zeusmaster/slavenode.go:134 +0x161
created by github.com/burke/zeus/go/zeusmaster._func_010
        /Users/burke/go/src/github.com/burke/zeus/go/zeusmaster/slavemonitor.go:46 +0x1f8
```
## Expected Behavior

The backtrace is sent to the client.

--- a/go/clienthandler/clienthandler.go
+++ b/go/clienthandler/clienthandler.go
@@ -74,17 +74,18 @@ func handleClientConnection(tree *processtree.ProcessTree, usock *unixsocket.Uso
 	defer clientFile.Close()
 
 	if err == nil && slaveNode.Error != "" {
-		// we can skip steps 3-5 as they deal with the command process we're not spawning.
-		// Write a fake pid (step 6)
-		usock.WriteMessage("0")
-		// Write the error message to the terminal
-		clientFile.Write([]byte(slaveNode.Error))
-		// Skip step 7, and write an exit code to the client (step 8)
-		usock.WriteMessage("1")
+		writeStacktrace(usock, slaveNode, clientFile)
 		return
 	}
 
 	commandUsock, err := bootNewCommand(slaveNode, command, err)
+	if err != nil {
+		// If a client connects while the command is just
+		// booting up, it actually makes it here - still
+		// expects a backtrace, of course.
+		writeStacktrace(usock, slaveNode, clientFile)
+		return
+	}
 	defer commandUsock.Close()
 
 	err = sendClientPidAndArgumentsToCommand(commandUsock, clientPid, arguments, err)
@@ -103,6 +104,16 @@ func handleClientConnection(tree *processtree.ProcessTree, usock *unixsocket.Uso
 		slog.Error(err)
 	}
 	// Done! Hooray!
+}
+
+func writeStacktrace(usock *unixsocket.Usock, slaveNode *processtree.SlaveNode, clientFile *os.File) {
+	// Fake process ID / output / error codes:
+	// Write a fake pid (step 6)
+	usock.WriteMessage("0")
+	// Write the error message to the terminal
+	clientFile.Write([]byte(slaveNode.Error))
+	// Write a non-positive exit code to the client
+	usock.WriteMessage("1")
 }
 
 func receiveCommandArgumentsAndPid(usock *unixsocket.Usock, err error) (string, int, string, error) {
@@ -209,12 +220,16 @@ func bootNewCommand(slaveNode *processtree.SlaveNode, command string, err error)
 		return nil, err
 	}
 
-	request := &processtree.CommandRequest{command, make(chan *os.File)}
+	request := &processtree.CommandRequest{command, make(chan *processtree.CommandReply)}
 	slaveNode.RequestCommandBoot(request)
-	commandFile := <-request.Retchan // TODO: don't really want to wait indefinitely.
+	reply := <-request.Retchan // TODO: don't really want to wait indefinitely.
 	// defer commandFile.Close() // TODO: can't do this here anymore.
 
-	return unixsocket.NewUsockFromFile(commandFile)
+	if reply.State == processtree.SCrashed {
+		return nil, errors.New("Process has crashed")
+	}
+
+	return unixsocket.NewUsockFromFile(reply.File)
 }
 
 func sendTTYToCommand(commandUsock *unixsocket.Usock, clientFile *os.File, err error) error {

--- a/go/processtree/slavenode.go
+++ b/go/processtree/slavenode.go
@@ -41,9 +41,14 @@ type SlaveNode struct {
 	event chan bool
 }
 
+type CommandReply struct {
+	State string
+	File  *os.File
+}
+
 type CommandRequest struct {
 	Name    string
-	Retchan chan *os.File
+	Retchan chan *CommandReply
 }
 
 const (
@@ -149,7 +154,7 @@ func (s *SlaveNode) Run(monitor *SlaveMonitor) {
 }
 
 // These "doXState" functions are called when a SlaveNode enters a state. They are expected
-// to continue to execute until 
+// to continue to execute until
 
 // The "SWaiting" state represents the state where a Slave is currently
 // not running, and neither is its parent. Before we can start booting
@@ -282,8 +287,11 @@ func (s *SlaveNode) bootSlave(slave *SlaveNode) {
 // command dies super early, the entire slave pretty well deadlocks.
 // TODO: review this.
 func (s *SlaveNode) bootCommand(request *CommandRequest) {
+	if s.State == SCrashed {
+		request.Retchan <- &CommandReply{SCrashed, nil}
+		return
+	}
 	identifier := request.Name
-	// TODO: If crashed, do something different...
 	msg := messages.CreateSpawnCommandMessage(identifier)
 	_, err := s.socket.WriteMessage(msg)
 	if err != nil {
@@ -298,7 +306,7 @@ func (s *SlaveNode) bootCommand(request *CommandRequest) {
 	}
 	fileName := strconv.Itoa(rand.Int())
 	commandFile := unixsocket.FdToFile(commandFD, fileName)
-	request.Retchan <- commandFile
+	request.Retchan <- &CommandReply{s.State, commandFile}
 }
 
 func (s *SlaveNode) bootQueuedCommandsAndSlaves() {

